pkg/ingester/user_state.go

```diff
@@ -263,10 +263,17 @@ func (u *userState) forSeriesMatching(ctx context.Context, allMatchers []*labels
 	}
 
 	level.Debug(log).Log("series", len(fps))
+	done := ctx.Done()
 
 	// fps is sorted, lock them in order to prevent deadlocks
 outer:
 	for _, fp := range fps {
+		select {
+		case <-done:
```

Comment:
My reading of the `Context` docs says you can just check `ctx.Err()` is non-nil and return if so, which ought to be lower-overhead than setting up a `select`. But who knows, really.